src/chainstate/burn/operations/peg_in.rs

```diff
@@ -0,0 +1,470 @@
+use stacks_common::codec::StacksMessageCodec;
+
+use crate::burnchains::BurnchainBlockHeader;
+use crate::burnchains::BurnchainTransaction;
+use crate::chainstate::burn::Opcodes;
+use crate::types::chainstate::StacksAddress;
+use crate::types::Address;
+
+use crate::chainstate::burn::operations::Error as OpError;
+use crate::chainstate::burn::operations::PegInOp;
+
+use crate::vm::errors::RuntimeErrorType as ClarityRuntimeError;
+use crate::vm::types::PrincipalData;
+use crate::vm::types::QualifiedContractIdentifier;
+use crate::vm::types::StandardPrincipalData;
+use crate::vm::ContractName;
+
+impl PegInOp {
+    pub fn from_tx(
+        block_header: &BurnchainBlockHeader,
+        tx: &BurnchainTransaction,
+    ) -> Result<Self, OpError> {
+        if tx.opcode() != Opcodes::PegIn as u8 {
+            warn!("Invalid tx: invalid opcode {}", tx.opcode());
+            return Err(OpError::InvalidInput);
+        }
+
+        let (amount, peg_wallet_address) =
+            if let Some(Some(recipient)) = tx.get_recipients().first() {
+                (recipient.amount, recipient.address.clone())
+            } else {
+                warn!("Invalid tx: Output 2 not provided");
```

Comment:
N.B. This can also fail if the output is not recognized (i.e. is non-standard).

Reply:
Thanks for clarifying. I've updated the error message.